core-util/SharedPointer.h

```diff
@@ -0,0 +1,233 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef __CORE_UTIL_SHAREDPOINTER_H__
+#define __CORE_UTIL_SHAREDPOINTER_H__
+
+#include "core-util/core-util.h"
+#include "mbed-alloc/ualloc.h"
+
+#include <stdint.h>
+#include <stddef.h>
+
+#ifndef NDEBUG
+#include <stdio.h>
+#define CORE_UTIL_SHAREDPOINTER_DEBUG(...) { printf(__VA_ARGS__); }
+#else
+#define CORE_UTIL_SHAREDPOINTER_DEBUG(...) /* nothing */
+#endif
+
+namespace mbed {
+namespace util {
+
+/** Shared pointer class.
+  *
+  * Similar to std::shared_ptr in C++11.
+  *
+  * Usage: SharedPointer<class> POINTER(new class())
+  *
+  * When POINTER is passed around by value the copy constructor and
+  * destructor counts the number of references to the original object.
+  * If the counter reaches zero, delete is called on the object pointed to.
+  *
+  * To avoid loops, "weak" references should be used by calling the original
+  * pointer directly through POINTER.get().
+  */
+
+template <class T>
+class SharedPointer {
+public:
+    /** Create empty SharedPointer not pointing to anything.
+      * Used for variable declaration.
+      */
+    SharedPointer(): pointer(NULL), counter(NULL) {
+        CORE_UTIL_SHAREDPOINTER_DEBUG("SP: %p [%p: %p]\r\n", this, pointer, counter);
+    }
+
+    /** Create new SharedPointer
+      * @param _pointer Pointer to take control over
+      */
+    SharedPointer(T* _pointer): pointer(_pointer) {
+        CORE_UTIL_ASSERT(pointer);
+
+        // allocate counter on the heap so it can be shared
+        UAllocTraits_t traits = {0};
+        counter = (uint32_t*) mbed_ualloc(sizeof(uint32_t), traits);
+
+        // initialize counter to 1
+        CORE_UTIL_ASSERT(counter);
+        *counter = 1;
+
+        CORE_UTIL_SHAREDPOINTER_DEBUG("SP: %p [%p: %p = %lu]\r\n", this, pointer, counter, *counter);
+    }
+
+    ~SharedPointer() {
+        decrementCounter();
+    }
+
+    /** Copy constructor
+      * Create new SharePointer from other SharedPointer by
+      * copying pointer to original object and pointer to counter.
+      * @param source object being copied from
+      */
+    SharedPointer(const SharedPointer& source): pointer(source.pointer), counter(source.counter) {
+        // increment reference counter
+        if (counter) {
+            (*counter)++;
+        }
+
+        CORE_UTIL_SHAREDPOINTER_DEBUG("SP&: %p = %p [%p: %p = %lu]\r\n", this, &source, pointer, counter, *counter);
+    }
+
+    /** Assignment operator.
+      * Cleanup previous reference and assign new pointer and counter.
+      * @param source object being assigned
+      */
+    SharedPointer operator=(const SharedPointer& source) {
+        if (*this != source) {
```

Comment:
`*this != source` will invoke one of the comparison operators, which might not be what you want. You're effectively checking against self assignment, so a better option is to check if the value of `this` is different from what you're assigning (`if (this != &source) ...`)